File: botasaurus_server/botasaurus_server/filters.py
import json
from casefy import snakecase, titlecase
import json

from .errors import JsonHTTPResponseWithMessage

class BaseFilter:
    def __init__(self, field, label=None):
        self.field = field
        self.label = label

        class_name = snakecase(self.__class__.__name__)
        self.id = f"{field}_{class_name}"

    def should_filter(self, filter_value):
        # General logic to determine if this filter should be applied
        return filter_value is not None

    def map_filter_value(self, filter_value):
        # Default implementation just returns the filter value as is.
        # Override this in subclasses as needed.
        return filter_value

    def filter(self, filter_value, data_value):
        # To be implemented by subclasses
        raise NotImplementedError

    def to_json(self):
        # Generate a dictionary that represents the JSON structure
        filter_json = {
            "id": self.id,
            "type": self.__class__.__name__,
            "label": self.label or self.get_label(),
        }
        return filter_json

    def get_label(self):
        raise NotImplementedError


class _NumericFilterBase(BaseFilter):
    def should_filter(self, filter_value):
        # Filter should be applied if filter_value is a number (int or float) and not None
        return filter_value is not None and isinstance(filter_value, (int, float))


class MinNumberInput(_NumericFilterBase):
    def filter(self, filter_value, data_value):
        # If data_value is None, it automatically fails the minimum number condition
        if isinstance(data_value, (int, float)):
            return data_value >= filter_value
        else:
            # Handle non-numeric data_value as per your requirements (e.g., log a warning or raise an error)
            return False

    def get_label(self):
        title_cased_field = titlecase(self.field)
        return f"Min {title_cased_field}"


class MaxNumberInput(_NumericFilterBase):
    def filter(self, filter_value, data_value):
        # If data_value is None, it automatically fails the maximum number condition
        if isinstance(data_value, (int, float)):
            return data_value <= filter_value
        else:
            # Handle non-numeric data_value as per your requirements
            return False

    def get_label(self):
        title_cased_field = titlecase(self.field)
        return f"Max {title_cased_field}"


class _Checkbox(BaseFilter):
    def should_filter(self, filter_value):
        return filter_value is True


class IsTrueCheckbox(_Checkbox):
    def filter(self, filter_value, data_value):
        return data_value is True

    def get_label(self):
        title_cased_field = titlecase(self.field)
        return f"Has {title_cased_field}"


class IsFalseCheckbox(_Checkbox):
    def filter(self, filter_value, data_value):
        return data_value is False

    def get_label(self):
        title_cased_field = titlecase(self.field)
        return f"Does Not Have {title_cased_field}"


class IsNullCheckbox(_Checkbox):
    def filter(self, filter_value, data_value):
        return data_value is None

    def get_label(self):
        title_cased_field = titlecase(self.field)
        return f"Does Not Have {title_cased_field}"


class IsNotNullCheckbox(_Checkbox):
    def filter(self, filter_value, data_value):
        return data_value is not None

    def get_label(self):
        title_cased_field = titlecase(self.field)
        return f"Has {title_cased_field}"


class IsTruthyCheckbox(_Checkbox):
    def filter(self, filter_value, data_value):
        return data_value

    def get_label(self):
        title_cased_field = titlecase(self.field)
        return f"Has {title_cased_field}"


class IsFalsyCheckbox(_Checkbox):
    def filter(self, filter_value, data_value):
        # Filter passes if data_value is falsy (evaluates to False in a boolean context)
        return not data_value

    def get_label(self):
        title_cased_field = titlecase(self.field)
        return f"Does Not Have {title_cased_field}"


class _DropdownFilterBase(BaseFilter):
    def __init__(self, field, options, label=None, case_insensitive=True):
        super().__init__(field, label)

        self.options = options if options else []
        self.case_insensitive = case_insensitive

        if self.options:
            self.validate_options(self.options)

    def validate_options(self, options):
        if not all("value" in option and "label" in option for option in options):
            raise ValueError("Each option must have 'value' and 'label' keys.")
        if not all(
            isinstance(option["value"], str) and isinstance(option["label"], str)
            for option in options
        ):
            raise TypeError("'value' and 'label' must be of type str.")
        if len(set(option["value"] for option in options)) != len(options):
            raise ValueError("'value' must be unique for each option.")

    def to_json(self):
        # Call the BaseFilter's to_json method and extend its functionality
        filter_json = super().to_json()

        # Add the options attribute specific to the _DropdownFilterBase class
        filter_json["options"] = self.options

        return filter_json


class SingleSelectDropdown(_DropdownFilterBase):
    def filter(self, filter_value, data_value):
        if isinstance(data_value, str):
            comparison_value = (
                data_value.lower() if self.case_insensitive else data_value
            )
            return comparison_value == filter_value
        elif isinstance(data_value, list):
            for item in data_value:
                if isinstance(item, str):  # Ensure the item is a string
                    # Apply case insensitivity if needed and compare directly
                    if (
                        item.lower() if self.case_insensitive else item
                    ) == filter_value:
                        return True
        return False

    def map_filter_value(self, filter_value):
        if self.case_insensitive and isinstance(filter_value, str):
            return filter_value.lower()
        return filter_value

    def should_filter(self, filter_value):
        # General logic to determine if this filter should be applied
        return isinstance(filter_value, str) and filter_value.strip()

    def get_label(self):
        title_cased_field = titlecase(self.field)
        return f"{title_cased_field} Is"


class BoolSelectDropdown(_DropdownFilterBase):

    def __init__(self, field,  label=None, prioritize_no=False):
        super().__init__(field, label)

        if prioritize_no:
            self.options =[{"value":"no","label":"No"},{"value":"yes","label":"Yes"},]
        else:
            self.options =[{"value":"yes","label":"Yes"},{"value":"no","label":"No"}]

        # todo remove
        self.validate_options(self.options)

    def filter(self, filter_value, data_value):
        if filter_value == "yes":
            if data_value:
                return True
            else:
                return False
        elif filter_value == "no":
            if not data_value:
                return True
            else:
                return False

        return False

    def should_filter(self, filter_value):
        return isinstance(filter_value, str) and filter_value.strip()

    def get_label(self):
        title_cased_field = titlecase(self.field)
        return f"Has {title_cased_field}"


class MultiSelectDropdown(_DropdownFilterBase):
    def filter(self, filter_value, data_value):
        if isinstance(data_value, str):
            comparison_value = (
                data_value.lower() if self.case_insensitive else data_value
            )
            return comparison_value in filter_value
        elif isinstance(data_value, list):
            # Directly iterate over data_value and check against filter_value
            for item in data_value:
                if isinstance(item, str):
                    # Apply case insensitivity if needed
                    item = item.lower() if self.case_insensitive else item
                    if item in filter_value:
                        return True
        return False

    def map_filter_value(self, filter_value):
        if not isinstance(filter_value, (set, list, tuple)):
            raise JsonHTTPResponseWithMessage(
                "filter_value must be a set, list, or tuple for MultiSelectDropdown"
            )

        if filter_value and not all(isinstance(val, str) for val in filter_value):
            raise JsonHTTPResponseWithMessage("filter_value must only contain strings")

        if self.case_insensitive:
            return set(val.lower() for val in filter_value if isinstance(val, str))
        else:
            return set(filter_value)

    def get_label(self):
        title_cased_field = titlecase(self.field)
        return f"{title_cased_field} Is One Of"


class SearchTextInput(BaseFilter):
    def filter(self, filter_value, data_value):
        # Filter passes if the filter_value (search term) is a substring of data_value
        # Assumes data_value is a string or can be cast to a string
        if isinstance(data_value, str):
            return filter_value in data_value.lower()
        return False

    def should_filter(self, filter_value):
        # General logic to determine if this filter should be applied
        return isinstance(filter_value, str) and filter_value.strip()

    def map_filter_value(self, filter_value):
        # Default implementation just returns the filter value as is.
        # Override this in subclasses as needed.
        return filter_value.strip().lower()

    def get_label(self):
        title_cased_field = titlecase(self.field)
        return f"Search {title_cased_field}"


def apply_filters(data_records, filters_data, available_filters):
    if not filters_data:
        return data_records

    # Determine which filters are applicable
    applicable_filters = [
        filter_obj
        for filter_obj in available_filters
        if filter_obj.should_filter(filters_data.get(filter_obj.id))
    ]

    # Update filters_data by mapping each filter's value using its map_filter_value method
    for filter_obj in applicable_filters:
        original_filter_value = filters_data[filter_obj.id]
        # Use the map_filter_value method to transform the filter value
        transformed_filter_value = filter_obj.map_filter_value(original_filter_value)
        # Update the filter value in filters_data with the transformed value
        filters_data[filter_obj.id] = transformed_filter_value

    # Apply each filter to the data records
    filtered_records = []
    for record in data_records:
        passes_all_filters = True
        for filter_obj in applicable_filters:
            if filter_obj.field not in record:
                rep = json.dumps(record, indent=4)
                if len(rep) >= 1003:
                    rep = rep[:1000] + "..."
                raise JsonHTTPResponseWithMessage(
                    f"Filter field {filter_obj.field} not found in data record: {rep}."
                )
            if not filter_obj.filter(
                filters_data[filter_obj.id], record[filter_obj.field]
            ):
                passes_all_filters = False
                break

        if passes_all_filters:
            filtered_records.append(record)

    return filtered_records

File: js/botasaurus-server-js/src/filters.ts
import { snakeCase, capitalCase as titleCase } from 'change-case';
import { JsonHTTPResponseWithMessage } from './errors';
import { isNotNullish, isNullish } from './null-utils';
import { isEmpty } from './utils';

export abstract class BaseFilter {
  field: string;
  field_keys: string[];
  label: string | null;
  id: string;

  constructor(field: string, options?: { label?: string }) {
    this.field = field;
    this.field_keys = this.field.split(".");
    this.label = options?.label || null;

    const className = snakeCase(this.getClassName());
    this.id = `${field}_${className}`;
  }

  shouldFilter(filterValue: any): boolean {
    return isNotNullish(filterValue);
  }

  mapFilterValue(filterValue: any): any {
    return filterValue;
  }

  // Checks if the field exists in the given record (supports dot notation)
  hasFilterValue(record: Record<string, any>): boolean {

    if (this.field_keys.length == 1) {
      return this.field in record;
    }

    let current = record;

    for (const key of this.field_keys) {
      if (!(key in current)) {
        return false;
      }
      current = current[key];
    }

    return true;
  }

  getFilterValue(record: Record<string, any>): any {
    if (this.field_keys.length == 1) {
      return record[this.field]
    }

    let current = record;

    for (const key of this.field_keys) {
      current = current[key];
    }

    return current;
  }


  abstract filter(filterValue: any, dataValue: any): boolean;

  toJson(): Record<string, any> {
    return {
      id: this.id,
      type: this.getClassName(),
      label: this.label || this.getLabel(),
    };
  }

  abstract getLabel(): string;

  abstract getClassName(): string;
}

abstract class _NumericFilterBase extends BaseFilter {
  override shouldFilter(filterValue: any): boolean {
    return isNotNullish(filterValue) && typeof filterValue === 'number';
  }
}

export class MinNumberInput extends _NumericFilterBase {
  filter(filterValue: number, dataValue: any): boolean {
    if (typeof dataValue === 'number') {
      return dataValue >= filterValue;
    } else {
      return false;
    }
  }

  getLabel(): string {
    const titleCasedField = titleCase(this.field);
    return `Min ${titleCasedField}`;
  }

  getClassName(): string {
    return 'MinNumberInput';
  }
}

export class MaxNumberInput extends _NumericFilterBase {
  filter(filterValue: number, dataValue: any): boolean {
    if (typeof dataValue === 'number') {
      return dataValue <= filterValue;
    } else {
      return false;
    }
  }

  getLabel(): string {
    const titleCasedField = titleCase(this.field);
    return `Max ${titleCasedField}`;
  }

  getClassName(): string {
    return 'MaxNumberInput';
  }
}

abstract class _Checkbox extends BaseFilter {
  override shouldFilter(filterValue: any): boolean {
    return filterValue === true;
  }
}

export class IsTrueCheckbox extends _Checkbox {
  filter(_: boolean, dataValue: any): boolean {
    return dataValue === true;
  }

  getLabel(): string {
    const titleCasedField = titleCase(this.field);
    return `Has ${titleCasedField}`;
  }

  getClassName(): string {
    return 'IsTrueCheckbox';
  }
}

export class IsFalseCheckbox extends _Checkbox {
  filter(_: boolean, dataValue: any): boolean {
    return dataValue === false;
  }

  getLabel(): string {
    const titleCasedField = titleCase(this.field);
    return `Does Not Have ${titleCasedField}`;
  }

  getClassName(): string {
    return 'IsFalseCheckbox';
  }
}

export class IsNullCheckbox extends _Checkbox {
  filter(_: boolean, dataValue: any): boolean {
    return isNullish(dataValue);
  }

  getLabel(): string {
    const titleCasedField = titleCase(this.field);
    return `Does Not Have ${titleCasedField}`;
  }

  getClassName(): string {
    return 'IsNullCheckbox';
  }
}

export class IsNotNullCheckbox extends _Checkbox {
  filter(_: boolean, dataValue: any): boolean {
    return isNotNullish(dataValue);
  }

  getLabel(): string {
    const titleCasedField = titleCase(this.field);
    return `Has ${titleCasedField}`;
  }

  getClassName(): string {
    return 'IsNotNullCheckbox';
  }
}

export class IsTruthyCheckbox extends _Checkbox {
  filter(_: boolean, dataValue: any): boolean {
    return !!dataValue;
  }

  getLabel(): string {
    const titleCasedField = titleCase(this.field);
    return `Has ${titleCasedField}`;
  }

  getClassName(): string {
    return 'IsTruthyCheckbox';
  }
}

export class IsFalsyCheckbox extends _Checkbox {
  filter(_: boolean, dataValue: any): boolean {
    return !dataValue;
  }

  getLabel(): string {
    const titleCasedField = titleCase(this.field);
    return `Does Not Have ${titleCasedField}`;
  }

  getClassName(): string {
    return 'IsFalsyCheckbox';
  }
}

interface Option {
  value: string;
  label: string;
}

abstract class _DropdownFilterBase extends BaseFilter {
  options: Option[];
  caseInsensitive: boolean;

  constructor(
    field: string,
    options: Option[],
    dropdownOptions?: {
      label?: string;
      caseInsensitive?: boolean;
    }
  ) {
    super(field, { label: dropdownOptions?.label });
    this.options = options || [];
    this.caseInsensitive = dropdownOptions?.caseInsensitive ?? true;
    if (this.options.length) {
      this.validateOptions(this.options);
    }
  }

  validateOptions(options: Option[]): void {
    if (!options.every((option) => 'value' in option && 'label' in option)) {
      throw new Error("Each option must have 'value' and 'label' keys.");
    }
    if (
      !options.every(
        (option) => typeof option.value === 'string' && typeof option.label === 'string'
      )
    ) {
      throw new TypeError("'value' and 'label' must be of type string.");
    }
    if (new Set(options.map((option) => option.value)).size !== options.length) {
      throw new Error("'value' must be unique for each option.");
    }
  }

  override toJson(): Record<string, any> {
    const filterJson = super.toJson();
    filterJson.options = this.options;
    return filterJson;
  }
}

export class SingleSelectDropdown extends _DropdownFilterBase {
  filter(filterValue: string, dataValue: any): boolean {
    if (typeof dataValue === 'string') {
      const comparisonValue = this.caseInsensitive ? dataValue.toLowerCase() : dataValue;
      return comparisonValue === filterValue;
    } else if (Array.isArray(dataValue)) {
      for (const item of dataValue) {
        if (typeof item === 'string') {
          if ((this.caseInsensitive ? item.toLowerCase() : item) === filterValue) {
            return true;
          }
        }
      }
    }
    return false;
  }

  override mapFilterValue(filterValue: string): string {
    return this.caseInsensitive ? filterValue.toLowerCase() : filterValue;
  }

  override shouldFilter(filterValue: any): boolean {
    return typeof filterValue === 'string' && filterValue.trim() !== '';
  }

  getLabel(): string {
    const titleCasedField = titleCase(this.field);
    return `${titleCasedField} Is`;
  }

  getClassName(): string {
    return 'SingleSelectDropdown';
  }
}

export class BoolSelectDropdown extends _DropdownFilterBase {
  invertFilter: boolean;

  constructor(
    field: string,
    boolOptions?: {
      label?: string;
      prioritizeNo?: boolean;
      invertFilter?: boolean;
    }
  ) {
    const options = boolOptions?.prioritizeNo
      ? [
          { value: 'no', label: 'No' },
          { value: 'yes', label: 'Yes' },
        ]
      : [
          { value: 'yes', label: 'Yes' },
          { value: 'no', label: 'No' },
        ];
    super(field, options, { label: boolOptions?.label });
    this.invertFilter = boolOptions?.invertFilter ?? false;
  }

  filter(filterValue: string, dataValue: any): boolean {
    if (this.invertFilter) {
      if (filterValue === 'yes') {
        return !dataValue;
      } else if (filterValue === 'no') {
        return !!dataValue;
      }
    } else {
      if (filterValue === 'yes') {
        return !!dataValue;
      } else if (filterValue === 'no') {
        return !dataValue;
      }
    }
    return false;
  }

  override shouldFilter(filterValue: any): boolean {
    return typeof filterValue === 'string' && filterValue.trim() !== '';
  }

  getLabel(): string {
    const titleCasedField = titleCase(this.field);
    return `Has ${titleCasedField}`;
  }

  getClassName(): string {
    return 'BoolSelectDropdown';
  }
}

export class MultiSelectDropdown extends _DropdownFilterBase {
  filter(filterValue: Set<any>, dataValue: any): boolean {
    if (typeof dataValue === 'string') {
      const comparisonValue = this.caseInsensitive ? dataValue.toLowerCase() : dataValue;
      return filterValue.has(comparisonValue);
    } else if (Array.isArray(dataValue)) {
      for (const item of dataValue) {
        if (typeof item === 'string') {
          const comparisonItem = this.caseInsensitive ? item.toLowerCase() : item;
          if (filterValue.has(comparisonItem)) {
            return true;
          }
        }
      }
    }
    return false;
  }

  override mapFilterValue(filterValue: any): Set<string> {
    if (!Array.isArray(filterValue)) {
      throw new JsonHTTPResponseWithMessage('filterValue must be an array');
    }

    if (filterValue.length > 0 && !filterValue.every((val: any) => typeof val === 'string')) {
      throw new JsonHTTPResponseWithMessage('filterValue must only contain strings');
    }

    return new Set(
      this.caseInsensitive ? filterValue.map((val: string) => val.toLowerCase()) : filterValue
    );
  }

  getLabel(): string {
    const titleCasedField = titleCase(this.field);
    return `${titleCasedField} Is One Of`;
  }

  getClassName(): string {
    return 'MultiSelectDropdown';
  }
}

export class SearchTextInput extends BaseFilter {
  filter(filterValue: string, dataValue: any): boolean {
    if (typeof dataValue === 'string') {
      return dataValue.toLowerCase().includes(filterValue);
    }
    return false;
  }

  override shouldFilter(filterValue: any): boolean {
    return typeof filterValue === 'string' && filterValue.trim() !== '';
  }

  override mapFilterValue(filterValue: string): string {
    return filterValue.trim().toLowerCase();
  }

  getLabel(): string {
    const titleCasedField = titleCase(this.field);
    return `Search ${titleCasedField}`;
  }

  getClassName(): string {
    return 'SearchTextInput';
  }
}
export function applyFiltersInPlace(
  dataRecords: any[],
  filtersData: Record<string, any>,
  availableFilters: BaseFilter[]
): Record<string, any>[] {
  if (isEmpty(filtersData)) {
    return dataRecords;
  }
  

  const applicableFilters = availableFilters.filter((filterObj) =>
    filterObj.shouldFilter(filtersData[filterObj.id])
  );

  for (const filterObj of applicableFilters) {
    const originalFilterValue = filtersData[filterObj.id];
    const transformedFilterValue = filterObj.mapFilterValue(originalFilterValue);
    filtersData[filterObj.id] = transformedFilterValue;
  }

  const filteredRecords: Record<string, any>[] = [];
  for (let i = 0; i < dataRecords.length; i++) {
    const record = dataRecords[i];
    let passesAllFilters = true;
    for (const filterObj of applicableFilters) {
      if (!(filterObj.hasFilterValue(record))) {
        const rep = JSON.stringify(record, null, 4);
        throw new JsonHTTPResponseWithMessage(
          `Filter field ${filterObj.field} not found in data record: ${
            rep.length >= 1003 ? rep.slice(0, 1000) + '...' : rep
          }.`
        );
      }
      if (!filterObj.filter(filtersData[filterObj.id], filterObj.getFilterValue(record))) {
        passesAllFilters = false;
        break;
      }
    }

    if (passesAllFilters) {
      filteredRecords.push(record);
    }

    // Set the item to null to save space
    dataRecords[i] = null;
  }

  return filteredRecords;
}